File: pythonbible/pythonbible/bible/bible.py
"""Contains the Bible class."""

from __future__ import annotations

from functools import lru_cache
from typing import TYPE_CHECKING

from pythonbible.bible.errors import VersionMissingBookError
from pythonbible.bible.errors import VersionMissingChapterError
from pythonbible.bible.errors import VersionMissingVerseError

if TYPE_CHECKING:
    from pythonbible.books import Book
    from pythonbible.versions import Version


class Bible:
    """The Bible class.

    The Bible class contains the scripture content for a version and format along with
    the functionality necessary to get the scripture content for a verse or range of
    verses.

    :param version: The version of the Bible.
    :type version: Version
    :param scripture_content: The scripture content for the Bible.
    :type scripture_content: str
    :param verse_start_indices: The start indices for each verse.
    :type verse_start_indices: dict[int, int]
    :param verse_end_indices: The end indices for each verse.
    :type verse_end_indices: dict[int, int]
    :param max_verses: The maximum verses for each book and chapter.
    :type max_verses: dict[Book, dict[int, int]]
    :param short_titles: The short titles for each book.
    :type short_titles: dict[Book, str]
    :param long_titles: The long titles for each book.
    :type long_titles: dict[Book, str]
    :param is_html: Whether the scripture content is HTML.
    :type is_html: bool
    """

    version: Version
    scripture_content: str
    verse_start_indices: dict[int, int]
    verse_end_indices: dict[int, int]
    max_verses: dict[Book, dict[int, int]]
    short_titles: dict[Book, str]
    long_titles: dict[Book, str]
    is_html: bool

    def __init__(
        self: Bible,
        version: Version,
        scripture_content: str,
        verse_start_indices: dict[int, int],
        verse_end_indices: dict[int, int],
        max_verses: dict[Book, dict[int, int]],
        short_titles: dict[Book, str],
        long_titles: dict[Book, str],
        is_html: bool = False,
    ) -> None:
        """Initialize a Bible object.

        :param version: The version of the Bible.
        :type version: Version
        :param scripture_content: The scripture content for the Bible.
        :type scripture_content: str
        :param verse_start_indices: The start indices for each verse.
        :type verse_start_indices: dict[int, int]
        :param verse_end_indices: The end indices for each verse.
        :type verse_end_indices: dict[int, int]
        :param max_verses: The maximum verses for each book and chapter.
        :type max_verses: dict[Book, dict[int, int]]
        :param short_titles: The short titles for each book.
        :param short_titles: The short titles for each book.
        :param long_titles: The long titles for each book.
        :type long_titles: dict[Book, str]
        :param is_html: Whether the scripture content is HTML.
        :type is_html: bool
        """
        self.version = version
        self.scripture_content = scripture_content
        self.verse_start_indices = verse_start_indices
        self.verse_end_indices = verse_end_indices
        self.max_verses = max_verses
        self.short_titles = short_titles
        self.long_titles = long_titles
        self.is_html = is_html

    def get_scripture(
        self: Bible,
        start_verse_id: int,
        end_verse_id: int | None = None,
    ) -> str:
        """Get the scripture content for the given verse ID or range of verse IDs.

        :param start_verse_id: The starting verse ID.
        :type start_verse_id: int
        :param end_verse_id: The ending verse ID.
        :type end_verse_id: int | None
        :return: The scripture content for the given verse ID or range of verse IDs.
        :rtype: str
        :raises VersionMissingVerseError: if a verse ID is not valid for the version
        """
        if not self.is_valid_verse_id(start_verse_id):
            raise VersionMissingVerseError(self.version, start_verse_id)

        if end_verse_id and not self.is_valid_verse_id(end_verse_id):
            raise VersionMissingVerseError(self.version, end_verse_id)

        end_verse_id = end_verse_id or start_verse_id
        start_index, end_index = self._get_start_and_end_indices(
            start_verse_id,
            end_verse_id,
        )

        return _clean(self.scripture_content[start_index:end_index], self.is_html)

    def _get_start_and_end_indices(
        self: Bible,
        start_verse_id: int,
        end_verse_id: int,
    ) -> tuple[int, int]:
        start_index = self.verse_start_indices.get(start_verse_id, -1)
        end_index = self.verse_end_indices.get(end_verse_id, -1)
        return start_index, end_index

    def is_valid_verse_id(self, verse_id: int) -> bool:
        """Check if the given verse ID is valid in this Bible version.

        :param verse_id: a verse id
        :type verse_id: int
        :return: True if the verse_id is valid; otherwise, False.
        :rtype: bool
        """
        return verse_id in self.verse_start_indices

    def get_number_of_chapters(self, book: Book) -> int:
        """Get the number of chapters in the given book.

        :param book: a book of the Bible
        :type book: Book
        :return: the number of chapters in the given book
        :rtype: int
        :raises VersionMissingBookError: if the book is not valid for the version
        """
        if chapters := self.max_verses.get(book):
            return len(chapters)

        raise VersionMissingBookError(self.version, book)

    def get_number_of_verses(self, book: Book, chapter: int) -> int:
        """Get the number of verses in the given book and chapter.

        :param book: a book of the Bible
        :type book: Book
        :param chapter: a chapter number
        :type chapter: int
        :return: the number of verses in the given book and chapter
        :rtype: int
        :raises VersionMissingBookError: if the book is not valid for the version
        :raises VersionMissingChapterError: if the chapter is not valid for the book
        """
        chapters = self.max_verses.get(book)

        if not chapters:
            raise VersionMissingBookError(self.version, book)

        if chapter not in chapters:
            raise VersionMissingChapterError(self.version, book, chapter)

        return chapters.get(chapter, -1)

    def get_verse_ids(self) -> tuple[int, ...]:
        """Get all verse IDs in this Bible version.

        :return: A tuple of all verse IDs
        :rtype: tuple[int, ...]
        """
        return tuple(self.verse_start_indices.keys())


@lru_cache()
def _clean(scripture_content: str, is_html: bool) -> str:
    cleaned_content: str = scripture_content.strip()
    return clean_html(cleaned_content) if is_html else cleaned_content


@lru_cache()
def clean_html(scripture_content: str) -> str:
    if not scripture_content or scripture_content in {"</p><p>", "<p></p>"}:
        return ""

    cleaned_content: str = scripture_content

    if cleaned_content.endswith("<p>"):
        cleaned_content = cleaned_content[:-3]

    if not cleaned_content.startswith("<p>"):
        cleaned_content = f"<p>{cleaned_content}"

    if not cleaned_content.endswith("</p>"):
        cleaned_content = f"{cleaned_content}</p>"

    return "" if cleaned_content == "<p></p>" else cleaned_content

File: rustbible/src/bible.rs
use std::collections::HashMap;
use std::fmt::Debug;

use crate::versions::Version;
use crate::books::BookItem;

/// Errors analogous to the Python `VersionMissing*Error` exceptions.
#[derive(Debug)]
pub enum BibleError {
    VersionMissingBook { version: Version, book: BookItem },
    VersionMissingChapter { version: Version, book: BookItem, chapter: i32 },
    VersionMissingVerse { version: Version, verse_id: i32 },
}

impl std::fmt::Display for BibleError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            BibleError::VersionMissingBook { version, book } => {
                write!(f, "version {:?} missing book {:?}", version, book)
            }
            BibleError::VersionMissingChapter { version, book, chapter } => {
                write!(f, "version {:?} missing chapter {} in book {:?}", version, chapter, book)
            }
            BibleError::VersionMissingVerse { version, verse_id } => {
                write!(f, "version {:?} missing verse id {}", version, verse_id)
            }
        }
    }
}

impl std::error::Error for BibleError {}

/// The Bible struct containing scripture content and indices.
///
/// Uses concrete `Version` and `Book` types from the crate.
#[derive(Clone, Debug)]
pub struct Bible {
    pub version: Version,
    pub scripture_content: String,
    pub verse_start_indices: HashMap<i32, usize>,
    pub verse_end_indices: HashMap<i32, usize>,
    pub max_verses: HashMap<u8, HashMap<i32, i32>>,
    pub short_titles: HashMap<u8, String>,
    pub long_titles: HashMap<u8, String>,
    pub is_html: bool,
}

impl Bible {
    pub fn new(
        version: Version,
        scripture_content: String,
        verse_start_indices: HashMap<i32, usize>,
        verse_end_indices: HashMap<i32, usize>,
        max_verses: HashMap<u8, HashMap<i32, i32>>,
        short_titles: HashMap<u8, String>,
        long_titles: HashMap<u8, String>,
        is_html: bool,
    ) -> Self {
        Self {
            version,
            scripture_content,
            verse_start_indices,
            verse_end_indices,
            max_verses,
            short_titles,
            long_titles,
            is_html,
        }
    }

    /// Get scripture for a single verse id or a range.
    pub fn get_scripture(
        &self,
        start_verse_id: i32,
        end_verse_id: Option<i32>,
    ) -> Result<String, BibleError> {
        if !self.is_valid_verse_id(start_verse_id) {
            return Err(BibleError::VersionMissingVerse {
                version: self.version.clone(),
                verse_id: start_verse_id,
            });
        }

        if let Some(ev) = end_verse_id {
            if !self.is_valid_verse_id(ev) {
                return Err(BibleError::VersionMissingVerse {
                    version: self.version.clone(),
                    verse_id: ev,
                });
            }
        }

        let end_id = end_verse_id.unwrap_or(start_verse_id);
        let (start_index, end_index) = self.get_start_and_end_indices(start_verse_id, end_id)?;

        // Ensure indices are in-range for slicing
        let content_len = self.scripture_content.len();
        let s = start_index.min(content_len);
        let e = end_index.min(content_len).max(s);

        let slice = &self.scripture_content[s..e];
        Ok(clean(slice, self.is_html))
    }

    fn get_start_and_end_indices(
        &self,
        start_verse_id: i32,
        end_verse_id: i32,
    ) -> Result<(usize, usize), BibleError> {
        let start_index = self.verse_start_indices.get(&start_verse_id);
        let end_index = self.verse_end_indices.get(&end_verse_id);

        match (start_index, end_index) {
            (Some(&s), Some(&e)) => Ok((s, e)),
            (None, _) => Err(BibleError::VersionMissingVerse {
                version: self.version.clone(),
                verse_id: start_verse_id,
            }),
            (_, None) => Err(BibleError::VersionMissingVerse {
                version: self.version.clone(),
                verse_id: end_verse_id,
            }),
        }
    }

    /// Check if a verse id exists.
    pub fn is_valid_verse_id(&self, verse_id: i32) -> bool {
        self.verse_start_indices.contains_key(&verse_id)
    }

    /// Get number of chapters for a book.
    pub fn get_number_of_chapters(&self, book: &BookItem) -> Result<usize, BibleError> {
        if let Some(chapters) = self.max_verses.get(&book.number) {
            Ok(chapters.len())
        } else {
            Err(BibleError::VersionMissingBook {
                version: self.version.clone(),
                book: (*book).clone(),
            })
        }
    }

    /// Get number of verses for a book and chapter.
    pub fn get_number_of_verses(&self, book: &BookItem, chapter: i32) -> Result<i32, BibleError> {
        if let Some(chapters) = self.max_verses.get(&book.number) {
            if !chapters.contains_key(&chapter) {
                return Err(BibleError::VersionMissingChapter {
                    version: self.version.clone(),
                    book: (*book).clone(),
                    chapter,
                });
            }
            Ok(*chapters.get(&chapter).unwrap_or(&-1))
        } else {
            Err(BibleError::VersionMissingBook {
                version: self.version.clone(),
                book: (*book).clone(),
            })
        }
    }

    /// Return all verse ids available.
    pub fn get_verse_ids(&self) -> Vec<i32> {
        self.verse_start_indices.keys().copied().collect()
    }
}

/// Trim and optionally clean HTML-like wrappers.
pub fn clean(scripture_content: &str, is_html: bool) -> String {
    let cleaned = scripture_content.trim();
    if is_html {
        clean_html(cleaned)
    } else {
        cleaned.to_string()
    }
}

/// Clean minimal HTML paragraph wrappers similar to the Python logic.
pub fn clean_html(scripture_content: &str) -> String {
    if scripture_content.is_empty()
        || scripture_content == "</p><p>"
        || scripture_content == "<p></p>"
    {
        return String::new();
    }

    let mut s = scripture_content.to_string();

    if s.ends_with("<p>") {
        let len = s.len();
        // remove last 3 chars
        s.truncate(len.saturating_sub(3));
    }

    if !s.starts_with("<p>") {
        s = format!("<p>{}", s);
    }

    if !s.ends_with("</p>") {
        s = format!("{}{}", s, "</p>");
    }

    if s == "<p></p>" {
        String::new()
    } else {
        s
    }
}
